### ISSUE TITLE
marathon & fx2 grain modules cause master and minion failure

### ISSUE THREAD

**Issue by cedwards** [saltstack/salt#47006]
### Description of Issue/Question
A couple of proxy minion grains are failing to load properly on 2018.3.0 (FreeBSD) causing both the master and minion to fail.

The offending modules are `grains/marathon.py` and `grains/fx2.py`.

I can provide the `sysutils/py-salt-2018.3.0` port & Makefile if someone else needs to replicate.

### Setup
`pkg install py36-salt`
`service salt_master start`
`service salt_minion start`

### Steps to Reproduce Issue
After starting the master/minion I see the following in the logs and the service does not work:

```
==> /var/log/salt/master <==
2018-04-10 16:42:43,610 [salt.loader                                                                              :763 ][CRITICAL][48291] Failed to load grains defined in grain file opt-1.fx2 in function <function fx2 at 0x80dfa5510>, error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 756, in grains
    ret = funcs[key]()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 90, in fx2
    return _grains()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 72, in _grains
    (username, password) = _find_credentials()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 41, in _find_credentials
    usernames.append(__pillar__['proxy'].get('admin_username', 'root'))
  File "/usr/local/lib/python3.6/site-packages/salt/utils/context.py", line 228, in __getitem__
    return self._dict()[key]
KeyError: 'proxy'
2018-04-10 16:42:43,613 [salt.loader                                                                              :763 ][CRITICAL][48291] Failed to load grains defined in grain file opt-1.marathon in function <function marathon at 0x80dfaeb70>, error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 756, in grains
    ret = funcs[key]()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/marathon.py", line 41, in marathon
    "{0}/v2/info".format(__opts__['proxy'].get(
KeyError: 'proxy'

==> /var/log/salt/minion <==
2018-04-10 16:42:48,701 [salt.loader                                                                              :763 ][CRITICAL][60649] Failed to load grains defined in grain file opt-1.fx2 in function <function fx2 at 0x80de467b8>, error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 756, in grains
    ret = funcs[key]()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 90, in fx2
    return _grains()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 72, in _grains
    (username, password) = _find_credentials()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 41, in _find_credentials
    usernames.append(__pillar__['proxy'].get('admin_username', 'root'))
  File "/usr/local/lib/python3.6/site-packages/salt/utils/context.py", line 228, in __getitem__
    return self._dict()[key]
KeyError: 'proxy'
2018-04-10 16:42:48,703 [salt.loader                                                                              :763 ][CRITICAL][60649] Failed to load grains defined in grain file opt-1.marathon in function <function marathon at 0x80de4de18>, error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 756, in grains
    ret = funcs[key]()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/marathon.py", line 41, in marathon
    "{0}/v2/info".format(__opts__['proxy'].get(
KeyError: 'proxy'
```

### Fix(?)
To fix these issues I have applied the following patches

```diff
--- marathon.py	2018-04-02 10:35:12.000000000 -0600
+++ marathon.py.new	2018-04-10 16:53:59.551719000 -0600
@@ -14,10 +14,10 @@


 def __virtual__():
-    if not salt.utils.platform.is_proxy() or 'proxy' not in __opts__:
-        return False
-    else:
+    if salt.utils.platform.is_proxy():
         return __virtualname__
+    else:
+        return False


 def kernel():
```
I haven't done enough code-diving yet, but I wonder if it's related to the `salt.utils.platform.is_proxy()` function. Perhaps it's not accounting for FreeBSD properly. /shrug.

### Versions Report
```
Salt Version:
           Salt: 2018.3.0

Dependency Versions:
           cffi: 1.11.5
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.5 (default, Apr 10 2018, 00:56:14)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 17.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.3

System Versions:
           dist:
         locale: US-ASCII
        machine: amd64
        release: 11.1-RELEASE
         system: FreeBSD
        version: Not Installed
```

**Comment by gtmanfred** [comment]
Yeah, that just looks like it is done backwards, i would have done.

```
    if salt.utils.platform.is_proxy() and 'proxy' in __opts__:
        return __virtualname__
    return False
```

Thanks for reporting.
Daniel

**Comment by Ch3LL** [comment]
@cedwards does @gtmanfred 's fix work for you?

**Comment by cedwards** [comment]
The suggested patch does not seem to work. The grains are still trying to load.

I'm working on this now. I'll submit something this evening.

**Comment by cedwards** [comment]
Upon further investigation there may be an issue with my `__pycache__`. I'm still troubleshooting.

**Comment by ari** [comment]
Can we get this fixed for 2018.3.1?

**Comment by gtmanfred** [comment]
@ari are you able to verify the fix above? That is all we are waiting on before I open a PR.

**Comment by ari** [comment]
Yes, it worked for marathon for me. FreeBSD 11.1, salt under python 3.6 using  gtmanfred's approach.

However I have a similar problem at /usr/local/lib/python3.6/site-packages/salt/grains/fx2.py. I did this:

```
def __virtual__():
    try:
        if salt.utils.platform.is_proxy() and 'proxy' in __opts__ and __opts__['proxy']['proxytype'] == 'fx2':
            return __virtualname__
    except KeyError:
        pass
    return False
```

But then I get another error:

```
10:42:45,855 [CRITICAL] Failed to load grains defined in grain file opt-1.fx2 in function <function fx2 at 0x8104d6840>, error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 756, in grains
    ret = funcs[key]()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 90, in fx2
    return _grains()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 72, in _grains
    (username, password) = _find_credentials()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 41, in _find_credentials
    usernames.append(__pillar__['proxy'].get('admin_username', 'root'))
  File "/usr/local/lib/python3.6/site-packages/salt/utils/context.py", line 228, in __getitem__
    return self._dict()[key]
KeyError: 'proxy'
10:42:45,856 [CRITICAL] Failed to load grains defined in grain file opt-1.location in function <function location at 0x8104d6950>, error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 756, in grains
    ret = funcs[key]()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 99, in location
    GRAINS_CACHE.update(_grains())
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 72, in _grains
    (username, password) = _find_credentials()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/fx2.py", line 41, in _find_credentials
    usernames.append(__pillar__['proxy'].get('admin_username', 'root'))
  File "/usr/local/lib/python3.6/site-packages/salt/utils/context.py", line 228, in __getitem__
    return self._dict()[key]
KeyError: 'proxy'
10:42:45,857 [CRITICAL] Failed to load grains defined in grain file opt-1.marathon in function <function marathon at 0x8104d8ea0>, error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 756, in grains
    ret = funcs[key]()
  File "/usr/local/lib/python3.6/site-packages/salt/grains/marathon.py", line 41, in marathon
    "{0}/v2/info".format(__opts__['proxy'].get(
KeyError: 'proxy'
```

**Comment by cedwards** [comment]
How much did you delete from the `__pycache__`? I've removed the marathon and fx2 grains cache files and don't see the issue.

Still troubleshooting why this would affect py36 like this. I don't see the error on py27.

**Comment by jhujhiti** [comment]
Something else is going on here. A number of grains are "proxy". So far, spelunking in the code has gotten me nowhere. The value does **not** seem to be coming from grains/core.py line 1382 where it would appear to, though...

Removing the fx2 and marathon entries from `__pycache__` did resolve the exceptions, but my sense is that it's not something that needs fixing if this `proxy` thing were behaving itself.

```
% sudo salt-call grains.item osrelease
local:
    ----------
    osrelease:
        proxy
% uname -r
11.1-RELEASE-p9
```

**Comment by jhujhiti** [comment]
I knew I would come to a realization right after posting...

My grains behave themselves if I remove `salt/grains/__pycache__/*opt-1*` and break as soon as I reinstall the package with those files in it.

**Comment by ari** [comment]
Maybe related #47365

**Comment by jhujhiti** [comment]
A few more observations today: with the eg., `__pycache__/fx2.cpython-36.opt-1.pyc` files in place, if I run `salt-call -l trace grains.item kernel`, I see it attempting to load a bunch of bogus grains:

```
[TRACE   ] Loading opt-1.fx2 grain
[TRACE   ] Loading opt-1.kernel grain
[TRACE   ] Loading opt-1.location grain
```

These things `fx2`, `kernel`, and `location` are actually functions inside `fx2.py`. This is the source of the "proxy" values. Strangely, I don't see this behavior until I have at least two "problem" modules with optimized bytecode in place. I've been using esxi and fx2 as my test cases. With zero or one such optimized modules, I see no trace messages at all regarding the modules in question, which is what I expect because the `__virtual__` function would correctly disable them. Is `opt-1` being treated as if it were a module?

**Comment by jhujhiti** [comment]
(Sorry for the spam -- this was my weekend project and I'm determined to solve it.)

I stepped through the loader code with a debugger until I found the bits that were changed by this commit:
https://github.com/saltstack/salt/commit/4951b9a60d9bac1fa32ec9934abe0789c01e50f8#diff-f6777cb7fca1276c97cd4b4a6b9f085a
Specifically, `f_noext = f_noext.replace(BIN_PRE_EXT, '')` is an attempt to ignore the extra bits in the filename of a bytecode file, but it only removes `cpython-36`, so we end up loading a module named `fx2.opt-1`.

I'm not sure how to properly fix that, but we can work around it by not building optimized bytecode in the port. Setting `PYDISTUTILS_INSTALLARGS=-c --prefix=${PREFIX}` in the Makefile and removing the line `%%PYTHON3%%%%PYTHON_SITELIBDIR%%/salt/__pycache__/_syspaths.cpython-%%PYTHON_SUFFIX%%.%%PYTHON_PYOEXTENSION%%` in `pkg-plist` is the workaround I've hacked into my own ports tree.

**Comment by cedwards** [comment]
Testing the fix...

**Comment by cedwards** [comment]
Initial findings look good.

**Comment by cedwards** [comment]
patch submitted to FreeBSD: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=227836

**Comment by Ch3LL** [comment]
@jhujhiti what you are describing related to issue https://github.com/saltstack/salt/issues/46924 which you can track. 

ping @gtmanfred seems the fix has been verified mind sending up a PR?

**Comment by gtmanfred** [comment]
Done

**Comment by jhujhiti** [comment]
These are all caused by the same thing -- this one plus https://github.com/saltstack/salt/issues/46924 and https://github.com/saltstack/salt/issues/47365

**Comment by dereckson** [comment]
On FreeBSD 11.1 and 11.2 machines, with 2018.3.1 (current ports version), I had the same issue.

After manually applying 205abe2 + clearing pycache, it solves the issue:

(1) no more mention of the issue in the logs when command runs with -ldebug

(2) Expected osrelease grain value:

```
$  salt-call --local grains.item osrelease
local:
    ----------
    osrelease:
        11.2
```

**Comment by cedwards** [comment]
This should now be resolved in `sysutils/py-salt` 2018.3.2 in the FreeBSD ports tree. Please verify.

**Comment by szandala** [comment]
Greetings,
I still have the problem on Alpine Linux (both 3.8 and edge)
`[CRITICAL] Failed to load grains defined in grain file opt-1.marathon in function <function marathon at 0x7f2fc7611c80>, error:
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/loader.py", line 757, in grains
    ret = funcs[key]()
  File "usr/lib/python3.6/site-packages/salt/grains/marathon.py", line 40, in marathon
    "{0}/v2/info".format(__opts__['proxy'].get(
KeyError: 'proxy'
`

`/ # salt -V
Salt Version:
           Salt: 2018.3.2
 
Dependency Versions:
           cffi: 1.10.0
       cherrypy: Not Installed
       dateutil: 2.7.3
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: 0.27.2
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.27.1
         Python: 3.6.6 (default, Oct 18 2018, 16:04:36)
   python-gnupg: Not Installed
         PyYAML: 4.1
          PyZMQ: 17.1.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.2
            ZMQ: 4.2.5
 
System Versions:
           dist:   
         locale: UTF-8
        machine: x86_64
        release: 4.15.10-041510-generic
         system: Linux
        version: Not Installed
 
`

**Comment by Ch3LL** [comment]
and what about version 3.3? I know there was this PR https://github.com/saltstack/salt/pull/48636 that was added to properly handle pycache files.

**Comment by dynek** [comment]
Having the same problem on 2018.3.3 on Alpine 3.8

**Comment by Ch3LL** [comment]
thanks for clarifying. can you open a new issue with the details as this issue was specific to freebsd and looks to be fixed.
